### DISCUSSION TITLE
Mutli Tenancy Set Session Variable

### DISCUSSION THREAD

**Discussion by nucle** [jOOQ/jOOQ#15247] [Q&A] [answered]
Hi, 
I try to set a session variable with JOOQ, but it does not work.
In my configuration, I added an ExecuteListener and modified the query.

```java
"SET \"svc.current_organization\" = '7998db24-e4da-4042-9fbf-c878ce0efcf6';" + ctx.sql())
```

When I execute the request, I get the following error:

```
Error while reading field: location_id, at JDBC index: 1
Cannot convert from 0 (class java.lang.Integer) to class java.util.UUID
```

When I execute: 

```java
dsl.execute("SET \"svc.current_organization\" = '7998db24-e4da-4042-9fbf-c878ce0efcf6';") 
```

before my query then it works.

I do not understand what I am doing wrong

````kotlin
class JOOQConfiguration(val tenant: TenantAwareContext) {

    fun create(): JooqCustomContext {
        LOGGER.debug("MyCustomConfigurationFactory: create")
        return object : JooqCustomContext {
            override fun apply(configuration: Configuration?) {
                configuration!!.set(InitialisingVariableListener())
                Log.debug("Config TenantAwareContext(tenant='${tenant.tenantId}')")
            }
        }
    }
}
````

My listener:
````kotlin
 class InitialisingVariableListener() : ExecuteListener {
    override fun renderEnd(ctx: ExecuteContext) {
        ctx.sql("SET \"svc_event.current_organization\" = '7998db24-e4da-4042-9fbf-c878ce0efcf6';" + ctx.sql())
        Log.debug("RENDER END: " + ctx.sql())
    }
}
````

BR,
Chris

**Comment by lukaseder** [comment, marked as answer]
Thanks for your message. I can reproduce this easily with:

```java
ctx.configuration().set(ExecuteListener.onRenderEnd(h -> h.sql("set \"svc.x\" = 'a';" + h.sql())));
ctx.select(inline("A")).fetch();
```

The result being:

```
Executing query          : select 'A' -- SQL rendered with a free trial version of jOOQ 3.19.0-SNAPSHOT               
-> with bind values      : select 'A'                                                                                 
Version                  : Database version is supported by dialect POSTGRES_15: 15beta2 (Debian 15~beta2-1.pgdg110+1)
Affected row(s)          : 0                                                                                          
Fetched result           : +----+                                                                                     
                         : |A   |                                                                                     
                         : +----+                                                                                     
                         : |0   |                                                                                     
                         : +----+                                                                                     
Fetched row(s)           : 1                                                                                          
```

So, what is really being fetched here is the update count of the statement you've added (`0`), instead of the actual value `A`. jOOQ occasionally does this kind of thing itself, internally, and has an internal `AbstractContext.skipUpdateCounts` flag to allow for specifying how many of these preceding update counts should be skipped.

You can set this flag from a `VisitListener`:

```java
ctx.configuration()
   .set(ExecuteListener.onRenderEnd(h -> h.sql("set \"svc.x\" = 'a';" + h.sql())))
   .set(VisitListener.onVisitStart(c -> {
            if (c.data("already-skipped", true) == null)
                c.context().skipUpdateCounts(1);
        }));

ctx.select(inline("A")).fetch();
```

That will override any previous value, though, so if jOOQ ever sets this flag internally, you'd have to make sure:

- That jOOQ's own expectation is met by incrementing the flag
- That the increment only happens once (`VisitListener` is invoked on every `QueryPart` in the query, so when incrementing, I'm using the `already-skipped` data flag)

It's probably worth exposing this flag to the `ExecuteListener` SPI via `ExecuteContext`:

- https://github.com/jOOQ/jOOQ/issues/15249

**Reply by lukaseder** [reply]
Note, in case you're using Stack Overflow, I think this would be a great question to ask there as well for better visibility, in case other users have the same question: https://stackoverflow.com/questions/tagged/jooq. I would then repeat my answer there.

**Reply by lukaseder** [reply]
I got my initial example wrong. The flag is always incremented, so calling `skipUpdateCounts(1)` really means "increment the number of skipped upgrade counts by 1", not "set it to 1".

**Reply by nucle** [reply]
Hi Lukas,
Thank you very much for your response.
In the evening, I will create a new question on Stack Overflow.

BR,
Chris

**Comment by nucle** [comment]
Hi @lukaseder,
here the link to Stackoverflow:

https://stackoverflow.com/questions/76540584/jooq-mutli-tenancy-with-set-session-local-database-variable

BR,
Chris

**Reply by lukaseder** [reply]
Thanks a lot. Will reply soon
